File: src/model/LessonModel.ts
import { Timestamp } from "firebase/firestore";

export interface ILessonModel {
  id: string;
  title: string;
  description: string;
  estimatedReadTime: number;
  levelNumber: number;
  pdfUrl: string;
  accessTier: string;
  isActive: boolean;
  publishedAt?: Date | null;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  levelId: string;
}

export class LessonModel implements ILessonModel {
  id: string;
  title: string;
  description: string;
  estimatedReadTime: number;
  levelNumber: number;
  pdfUrl: string;
  accessTier: string;
  isActive: boolean;
  publishedAt?: Date | null;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  levelId: string;

  constructor(data: ILessonModel) {
    this.id = data.id;
    this.title = data.title;
    this.description = data.description;
    this.estimatedReadTime = data.estimatedReadTime;
    this.levelNumber = data.levelNumber;
    this.pdfUrl = data.pdfUrl;
    this.accessTier = data.accessTier;
    this.isActive = data.isActive;
    this.publishedAt = data.publishedAt ?? null;
    this.sortOrder = data.sortOrder;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.createdBy = data.createdBy;
    this.levelId = data.levelId;
  }

  /** Factory: create a new LessonModel instance (Dart: LessonModel.create) */
  static create(params: {
    id: string;
    title: string;
    description: string;
    estimatedReadTime: number;
    levelNumber: number;
    pdfUrl: string;
    accessTier?: string;
    isActive?: boolean;
    publishedAt?: Date | null;
    sortOrder?: number;
    createdBy?: string;
    levelId?: string;
  }): LessonModel {
    const now = new Date();
    return new LessonModel({
      id: params.id,
      title: params.title,
      description: params.description,
      estimatedReadTime: params.estimatedReadTime,
      levelNumber: params.levelNumber,
      pdfUrl: params.pdfUrl,
      accessTier: params.accessTier ?? "free",
      isActive: params.isActive ?? true,
      publishedAt: params.publishedAt ?? null,
      sortOrder: params.sortOrder ?? 0,
      createdAt: now,
      updatedAt: now,
      createdBy: params.createdBy ?? "admin",
      levelId: params.levelId ?? "",
    });
  }

  /** Convert to Firestore JSON (Dart: toFirestore) */
  toFirestore(): Record<string, any> {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      estimatedReadTime: this.estimatedReadTime,
      levelNumber: this.levelNumber,
      pdfUrl: this.pdfUrl,
      accessTier: this.accessTier,
      isActive: this.isActive,
      publishedAt: this.publishedAt
        ? Timestamp.fromDate(this.publishedAt)
        : null,
      sortOrder: this.sortOrder,
      createdAt: Timestamp.fromDate(this.createdAt),
      updatedAt: Timestamp.fromDate(this.updatedAt),
      createdBy: this.createdBy,
      levelId: this.levelId,
    };
  }

  /** Convert Firestore doc → LessonModel (Dart: LessonModel.fromFirestore) */
  static fromFirestore(json: Record<string, any>): LessonModel {
    return new LessonModel({
      id: json.id ?? "",
      title: json.title ?? "",
      description: json.description ?? "",
      estimatedReadTime: json.estimatedReadTime ?? 0,
      levelNumber: json.levelNumber ?? 1,
      pdfUrl: json.pdfUrl ?? "",
      accessTier: json.accessTier ?? "free",
      isActive: json.isActive ?? true,
      publishedAt: LessonModel._toDateTime(json.publishedAt),
      sortOrder: json.sortOrder ?? 0,
      createdAt: LessonModel._toDateTime(json.createdAt),
      updatedAt: LessonModel._toDateTime(json.updatedAt),
      createdBy: json.createdBy ?? "admin",
      levelId: json.levelId ?? "",
    });
  }

  /** Create a copy with modified fields (Dart: copyWith) */
  copyWith(updates: Partial<ILessonModel>): LessonModel {
    return new LessonModel({
      ...this,
      ...updates,
    });
  }

  /** Convert Firestore Timestamp or string → Date */
  private static _toDateTime(value: any): Date | null {
    if (!value) return null;
    if (value instanceof Timestamp) return value.toDate();
    if (typeof value === "string") return new Date(value);
    if (value instanceof Date) return value;
    return new Date(); // fallback like Dart's DateTime.now()
  }
}

File: Models/lesson models/lesson_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class LessonModel {
  final String id;
  final String title;
  final String description;
  final int estimatedReadTime;
  final int levelNumber;
  final String pdfUrl;
  final String accessTier;
  final bool isActive;
  final DateTime? publishedAt;
  final int sortOrder;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String createdBy;

  const LessonModel({
    required this.id,
    required this.title,
    required this.description,
    required this.estimatedReadTime,
    required this.levelNumber,
    required this.pdfUrl,
    required this.accessTier,
    required this.isActive,
    this.publishedAt,
    required this.sortOrder,
    required this.createdAt,
    required this.updatedAt,
    required this.createdBy,
  });

  factory LessonModel.create({
    required String id,
    required String title,
    required String description,
    required int estimatedReadTime,
    required int levelNumber,
    required String pdfUrl,
    String accessTier = 'free',
    bool isActive = true,
    DateTime? publishedAt,
    int sortOrder = 0,
    String createdBy = 'admin',
  }) {
    final now = DateTime.now();
    return LessonModel(
      id: id,
      title: title,
      description: description,
      estimatedReadTime: estimatedReadTime,
      levelNumber: levelNumber,
      pdfUrl: pdfUrl,
      accessTier: accessTier,
      isActive: isActive,
      publishedAt: publishedAt,
      sortOrder: sortOrder,
      createdAt: now,
      updatedAt: now,
      createdBy: createdBy,
    );
  }

  Map<String, dynamic> toFirestore() => {
    'id': id,
    'title': title,
    'description': description,
    'estimatedReadTime': estimatedReadTime,
    'levelNumber': levelNumber,
    'pdfUrl': pdfUrl,
    'accessTier': accessTier,
    'isActive': isActive,
    'publishedAt': publishedAt == null
        ? null
        : Timestamp.fromDate(publishedAt!),
    'sortOrder': sortOrder,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
    'createdBy': createdBy,
  };

  factory LessonModel.fromFirestore(Map<String, dynamic> json) => LessonModel(
    id: json['id'] ?? '',
    title: json['title'] ?? '',
    description: json['description'] ?? '',
    estimatedReadTime: json['estimatedReadTime'] ?? 0,
    levelNumber: json['levelNumber'] ?? 1,
    pdfUrl: json['pdfUrl'] ?? '',
    accessTier: json['accessTier'] ?? 'free',
    isActive: json['isActive'] ?? true,
    publishedAt: _toDateTime(json['publishedAt']),
    sortOrder: json['sortOrder'] ?? 0,
    createdAt: _toDateTime(json['createdAt']),
    updatedAt: _toDateTime(json['updatedAt']),
    createdBy: json['createdBy'] ?? 'admin',
  );

  LessonModel copyWith({
    String? id,
    String? title,
    String? description,
    int? estimatedReadTime,
    int? levelNumber,
    String? pdfUrl,
    String? accessTier,
    bool? isActive,
    DateTime? publishedAt,
    int? sortOrder,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? createdBy,
  }) => LessonModel(
    id: id ?? this.id,
    title: title ?? this.title,
    description: description ?? this.description,
    estimatedReadTime: estimatedReadTime ?? this.estimatedReadTime,
    levelNumber: levelNumber ?? this.levelNumber,
    pdfUrl: pdfUrl ?? this.pdfUrl,
    accessTier: accessTier ?? this.accessTier,
    isActive: isActive ?? this.isActive,
    publishedAt: publishedAt ?? this.publishedAt,
    sortOrder: sortOrder ?? this.sortOrder,
    createdAt: createdAt ?? this.createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
    createdBy: createdBy ?? this.createdBy,
  );

  static DateTime _toDateTime(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is String) return DateTime.parse(v);
    return DateTime.now();
  }
}
